daemon/algod/api/server/v2/utils.go

```diff
@@ -0,0 +1,165 @@
+}
+
+func badRequest(ctx echo.Context, internal error, external string, log logging.Logger) error {
+	return returnError(ctx, http.StatusBadRequest, internal, external, log)
+}
+
+func internalError(ctx echo.Context, internal error, external string, log logging.Logger) error {
+	return returnError(ctx, http.StatusInternalServerError, internal, external, log)
+}
+
+func notFound(ctx echo.Context, internal error, external string, log logging.Logger) error {
+	return returnError(ctx, http.StatusNotFound, internal, external, log)
+}
+
+func addrOrNil(addr basics.Address) *string {
+	if addr.IsZero() {
+		return nil
+	}
+	ret := addr.String()
+	return &ret
+}
+
+func strOrNil(str string) *string {
+	if str == "" {
+		return nil
+	}
+	return &str
+}
+
+func numOrNil(num uint64) *uint64 {
+	if num == 0 {
+		return nil
+	}
+	return &num
+}
+
+func byteOrNil(data []byte) *[]byte {
+	if len(data) == 0 {
+		return nil
+	}
+	return &data
+}
+
+func computeAssetIndexInPayset(tx node.TxnWithStatus, txnCounter uint64, payset []transactions.SignedTxnWithAD) (aidx *uint64) {
+	// Compute transaction index in block
+	offset := -1
+	for idx, stxnib := range payset {
+		if tx.Txn.Txn.ID() == stxnib.Txn.ID() {
+			offset = idx
+			break
+		}
+	}
+
+	// Sanity check that txn was in fetched block
+	if offset < 0 {
+		return nil
+	}
+
+	// Count into block to get created asset index
+	idx := txnCounter - uint64(len(payset)) + uint64(offset) + 1
+	return &idx
+}
+
+// computeAssetIndexFromTxn returns the created asset index given a confirmed
+// transaction whose confirmation block is available in the ledger. Note that
+// 0 is an invalid asset index (they start at 1).
+func computeAssetIndexFromTxn(tx node.TxnWithStatus, l *data.Ledger) (aidx *uint64) {
+	// Must have ledger
+	if l == nil {
+		return nil
+	}
+	// Transaction must be confirmed
+	if tx.ConfirmedRound == 0 {
+		return nil
+	}
+	// Transaction must be AssetConfig transaction
+	if tx.Txn.Txn.AssetConfigTxnFields == (transactions.AssetConfigTxnFields{}) {
+		return nil
+	}
+	// Transaction must be creating an asset
+	if tx.Txn.Txn.AssetConfigTxnFields.ConfigAsset != 0 {
+		return nil
+	}
+
+	// Look up block where transaction was confirmed
+	blk, err := l.Block(tx.ConfirmedRound)
+	if err != nil {
+		return nil
+	}
+
+	payset, err := blk.DecodePaysetFlat()
+	if err != nil {
+		return nil
+	}
+
+	return computeAssetIndexInPayset(tx, blk.BlockHeader.TxnCounter, payset)
+}
+
+func getCodecHandle(formatPtr *string) (codec.Handle, error) {
+	format := "json"
+	if formatPtr != nil {
+		format = strings.ToLower(*formatPtr)
+	}
+
+	var handle codec.Handle = protocol.JSONHandle
+	if format == "json" {
+		handle = protocol.JSONHandle
+	} else if format == "msgpack" || format == "msgp" {
+		handle = protocol.CodecHandle
+	} else {
+		return nil, fmt.Errorf("invalid format: %s", format)
+	}
+
+	return handle, nil
+}
+
+func encode(handle codec.Handle, obj interface{}) (string, error) {
```

Comment:
what does this function do?
encode from what to what?

Reply:
It encodes to JSON or MessagePack depending on which codec handle is passed in.